### PR TITLE
Java baseball 손현경

### PR DESCRIPTION
**Pull request by shkisme**
야구 게임
소요시간 : 약 4~5시간 소요...
테스트 코드 만들기가 더 어려웠던 것 같습니다

### CODE REVIEW

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Constant.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package java_baseball.java_baseball;
+
+public class Constant {
```
Comment (gusah009) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Constant.java:3]: 각 상수값들을 모두 상수화 하는 것은 좋은 것 같습니다! 개인적인 생각으로는 상수를 모두 여기에 몰아 넣는 것 보다, 각 상수가 쓰이는 클래스에 선언하는 게 좋아보입니다.

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    private int strike;
+    private int ball;
+    private boolean gameState = true;
+    private boolean restartState = false;
+
+    public void generateRandomNum(){
+        List<Integer> randomList = new ArrayList<>();
+        Random random = new Random();
+        while(randomList.size() < Constant.LIST_SIZE){
+            int randomNum = random.nextInt(100) % Constant.BALL_MAX + Constant.BALL_MIN;
```
Comment (woo-chang) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java]: 범위에 대해 유연하게 랜덤값을 생성할 수 있는 좋은 방법이네요 👍

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Constant.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    public static final String STRIKE_STRING = "스트라이크 ";
+    public static final String NOTHING = "낫싱";
+    public static final String ALL_CORRECT = "3개의 숫자를 모두 맞히셨습니다! 게임 종료";
+    public static final String ASK_PLAYER_RESTART = "게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요.";
+
+    public static final String LENGTH_ERROR = "숫자 3개를 입력해 주세요.";
+    public static final String VALUE_ERROR = "1부터 9 사이의 숫자를 입력해 주세요.";
+    public static final String DUPNUM_ERROR = "숫자를 중복 없이 입력해 주세요.";
+    public static final String RESTART_ERROR = "숫자 1 또는 2를 입력해 주세요.";
+}
```
Comment (woo-chang) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Constant.java:7]: 열거 상수 패턴보다는 표현력과 타입 안전성을 보장하는 열거 타입을 사용하는 것이 좋아보입니다

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    }
+    public void isGameEnd(Player player){
+        if (strike == Constant.ALL_STRIKE){
+            gameClear(player);
+            gameState = false;
+        }
+    }
+    private void gameClear(Player player){
+        System.out.println(Constant.ALL_CORRECT);
+        int playersAnswer = askGameRestart(player);
```
Comment (woo-chang) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java]: 해당 int 값이 어떤 의미를 가지는지 불분명한 것 같습니다. 여기에도 열거 타입으로 표현력을 추가하는 방법이 좋아보이네요

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+public class Error {
+    public void validateLengthError(String[] inputArr){
+        if (inputArr.length != Constant.LIST_SIZE){
+            throw new IllegalArgumentException(Constant.LENGTH_ERROR);
+        }
+    }
+    public void validateNumValueError(List<Integer> playersNum){
+        for (int i= 0; i< Constant.LIST_SIZE;i++){
+            if (!(1<=playersNum.get(i) && playersNum.get(i) <= 9)){
```
Comment (woo-chang) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java]: 검증하는 부분을 메서드로 추출해도 좋은 것 같은데 어떻게 생각하시나요?

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+                throw new IllegalArgumentException(Constant.VALUE_ERROR);
+            }
+        }
+    }
+    public void validateDupNumError(List<Integer> playersNum){
+        int num1 = playersNum.get(Constant.FIRST_INDEX);
+        int num2 = playersNum.get(Constant.SECOND_INDEX);
+        int num3 = playersNum.get(Constant.THIRD_INDEX);
+        if (num1 == num2 || num1==num3 || num2 == num3){
```
Comment (woo-chang) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java]: `Set<Integer> sets = new Set<>(playersNum)`을 사용하면 간편히 `Set`으로 바꿀 수 있는 것으로 알고 있는데 같은 값을 가지지 못하는 Set의 특성을 이용해 Set 원소의 개수로 검증하는 방법은 어떨까요? 리스트로 값을 관리하고 있기에 이런 방법도 괜찮을 것 같다는 생각이 들었습니다.

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+            throw new IllegalArgumentException(Constant.DUPNUM_ERROR);
+        }
+    }
+    public void validateRestartError(int restartSignal){
+        if (restartSignal != 1 && restartSignal != 2){
```
Comment (woo-chang) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java]: 1과 2가 담고있는 의미가 불명확한 것 같습니다. 열거 타입의 중요성이 계속 강조되는 것 같네요 😂

File: shkisme/java_baseball/src/test/java/java_baseball/java_baseball/ErrorTest.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+public class ErrorTest {
+    @Test
+    @DisplayName("숫자를 3개보다 많이 입력한 경우")
+    public void lengthError(){
+        Error error = new Error();
+        String[] inputArr = {"4","5","2","1","7"};
+        IllegalArgumentException e = assertThrows(IllegalArgumentException.class,
+                () -> error.validateLengthError(inputArr));//예외가 발생해야 한다.
+
+        assertThat(e.getMessage()).isEqualTo("숫자 3개를 입력해 주세요.");
```
Comment (woo-chang) [inline shkisme/java_baseball/src/test/java/java_baseball/java_baseball/ErrorTest.java]: `String`을 바로 비교하게 되면 해당 문구가 바뀌게 되었을 때 전부 수정해야 하는 문제가 생길 것 같네요

File: shkisme/java_baseball/src/test/java/java_baseball/java_baseball/RandomNumTest.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    @DisplayName("랜덤 수 생성 테스트 1")
+    public void randomGenerate1(){
+        Computer computer = new Computer();
+
+        computer.generateRandomNum();
+        List<Integer> randomNumList = computer.getRandomNumList();
+        System.out.println(randomNumList.get(0));
+        System.out.println(randomNumList.get(1));
+        System.out.println(randomNumList.get(2));
+    }
```
Comment (woo-chang) [inline shkisme/java_baseball/src/test/java/java_baseball/java_baseball/RandomNumTest.java]: `Test`의 검증은 `assert` 문구가 포함되어야 한다고 생각하는데 단순 출력으로는 테스트가 부족할 것 같습니다.

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Game.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    public void gameStart(){
+        do{
+            play();
+        } while(computer.getRestartState());
+    }
+    private void play(){
+        computer.generateRandomNum();
+        do{
+            computer.askNum();
+            computer.gameStateSet();
```
Comment (gusah009) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Game.java]: 크게 중요한 건 아닐 수도 있는데, `askNum()`과 gameStateSet()` 순서를 바꾸는 것도 좋아보입니다. 

`gameStateSet()`이 게임을 시작했다는 뜻 같은데, 게임이 시작되고나서 숫자를 물어보는 게 순서상 맞지 않나 싶습니다.

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        List<Integer> randomList = new ArrayList<>();
+        Random random = new Random();
+        while(randomList.size() < Constant.LIST_SIZE){
+            int randomNum = random.nextInt(100) % Constant.BALL_MAX + Constant.BALL_MIN;
+            if (!randomList.contains(randomNum))
+                randomList.add(randomNum);
+        }
+        randomNumList = randomList;
+    }
+    public void gameStateSet(){
```
Comment (gusah009) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java]: `gameStateSet`보다 더 나은 메서드명이 없을까요? `gameStateSet()` 메서드만으론 `gameState`가 `true`로 setting 되는 지, `false`로 setting 되는 지 알 수 없을 것 같습니다.

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
+package java_baseball.java_baseball;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+
+public class Computer {
```
Comment (gusah009) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java:10]: `Computer`가 **두 가지 역할**을 하고 있는 것 같습니다.

1. **상대 플레이어**로서의 역할.
   - `randomNumList`를 가지고 있다는 뜻은 상대 플레이어로서의 역할을 수행하고 있는 것으로 보입니다.
2. **게임 중계자**의 역할
   - `strike`, `ball`, `gameState`, `restartState` 같은 게임 전반에 필요한 정보를 함께 가지고 있습니다.

`Computer` 클래스가 역할을 여러개 갖고 있는 것 같아 클래스를 두 개로 분리하면 더 좋을 것 같습니다.

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+            System.out.println(ball + Constant.BALL_STRING);
+        }
+        else if (strike > 0){
+            System.out.println(strike + Constant.STRIKE_STRING);
+        }
+        else{
+            System.out.println(Constant.NOTHING);
+        }
+    }
+    public void isGameEnd(Player player){
```
Comment (gusah009) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java]: `isGameEnd`는 의문문으로 관례상 `is~~`로 시작하는 메서드는 `boolean` 타입을 반환합니다. 관례상 그렇긴 하지만, 실제로 문맥상으로도 가독성이 높아지기 때문에 되도록이면 `boolean` 타입을 반환하도록 하면 좋을 것 같습니다.

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+        if (strike == Constant.ALL_STRIKE){
+            gameClear(player);
+            gameState = false;
```
Comment (gusah009) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java]: > isGameEnd는 의문문으로 관례상 is~~로 시작하는 메서드는 boolean 타입을 반환합니다. 관례상 그렇긴 하지만, 실제로 문맥상으로도 가독성이 높아지기 때문에 되도록이면 boolean 타입을 반환하도록 하면 좋을 것 같습니다.

추가로, 그렇게 되면 `gameState`를 사용하지 않고도 재시작 여부를 판단할 수 있습니다.

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Game.java [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+        this.computer = new Computer();
+        this.player = new Player();
+    }
+    public void gameStart(){
+        do{
+            play();
+        } while(computer.getRestartState());
+    }
+    private void play(){
+        computer.generateRandomNum();
```
Comment (gusah009) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Game.java]: `computer.generateRandomNum();`는 왜 `do while` 밖에 있나요?

\+ `gameStart()`의 `do while`문과 `play()`의 `do while`문의 하는 일이 겹치는 것 같습니다.

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Player.java [hunk 16]
```diff
@@ -0,0 +1,31 @@
+package java_baseball.java_baseball;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
+
+public class Player {
+    Scanner sc = new Scanner(System.in);
+    Error error = new Error();
+    public List<Integer> inputNum(){
```
Comment (gusah009) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Player.java]: CleanCode issue에서 얘기했던 내용대로, **메서드 내부적으로 띄워쓰기가 필요한 부분은 함수로 분리하는 게 좋아보입니다!**

File: shkisme/java_baseball/src/test/java/java_baseball/java_baseball/ErrorTest.java (hunk 8 above)
Comment (gusah009) [reply shkisme/java_baseball/src/test/java/java_baseball/java_baseball/ErrorTest.java]: 덧붙이자면, `Constant` 클래스로 상수값을 잘 분리해서 관리하고 계셨던 것 같은데, 이를 이용해보면 좋을 것 같습니다!

File: shkisme/java_baseball/src/test/java/java_baseball/java_baseball/ErrorTest.java [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+    }
+    @Test
+    @DisplayName("숫자가 1보다 작거나, 9보다 큰 경우")
```
Comment (gusah009) [inline shkisme/java_baseball/src/test/java/java_baseball/java_baseball/ErrorTest.java]: 테스트에서는 `1, 5, 10`에 대해 테스트하고 있는데, 테스트 내용이 _"1보다 작거나, 9보다 큰 경우"_ 이므로 edge case로 `0`이 포함된 테스트 케이스도 만들면 좋을 것 같습니다.

**Review by gusah009 — COMMENTED** [review]
훌륭하네요... 과거의 저랑 비교돼서 반성하게 됩니다... 👍👍👍

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+    }
+    private void ballCnt(List<Integer> playersNum){
+        int cnt = 0;
+        for (int i =0 ;i < 3; i++){
+            if (this.randomNumList.contains(playersNum.get(i)) && !(playersNum.get(i).equals(this.randomNumList.get(i))))
+                cnt+=1;
+        }
+        this.ball = cnt;
+    }
+    public void printScore(){
```
Comment (JeonghakLee) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java]: strikeCnt, ballCnt와 같은 메서드 명을 사용하니 strike count와 같은 명사로 느껴져서 어떤 일을 하는지 의미를 알기가 힘든 것 같습니다. 
동작을 표현하는 동사를 먼저 써서 `countStrike` 처럼 사용하는 것이 의미를 더 명확히 할 수 있을 것 같습니다. 
또 수를 세는 역할과 수를 저장하는 역할 2가지를 하고 있는 것 같아서 메서드를 분리하는 것이 좋아보입니다.

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+    }
+    private void gameRestartJudging(int playersAnswer){
+        if (playersAnswer == Constant.RESTART_BUTTON){
+            restartState = true;
+        }
+        else{
+            restartState = false;
+        }
+    }
+    public boolean getRestartState(){
```
Comment (JeonghakLee) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java]: 위에서와 마찬가지로 메서드명을 judgeGameRestart 처럼 동사를 포함하여 어떤 동작을 하는지 알 수 있게하는 것이 좋아 보입니다.
또 상태를 판단하는 역할과 판단하여 저장하는 역할 두가지를 하는 것 처럼 보여 메서드를 분리하는 것이 좋아 보입니다!

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java (hunk 14 above)
Comment (JeonghakLee) [reply shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java]: 저도 `gameStatusTest` 테스트 코드를 보면서 한번 더 느꼈습니다 gameState 필드를  사용하지 않아도 될 것 같습니다.

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java (hunk 5 above)
Comment (shkisme) [reply shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java]: 값을 비교하는 메서드를 따로 만들어서 수정했습니다! 이렇게 하니까 훨씬 깔끔해진거 같아요👍

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java (hunk 6 above)
Comment (shkisme) [reply shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java]: Set을 이용하여 원소의 개수로 검증하는 로직이 간단하고 좋아서 이렇게 수정하였습니다!

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java (hunk 7 above)
Comment (shkisme) [reply shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java]: RestartStatus 열거 타입을 추가하여 수정하였습니다!

File: shkisme/java_baseball/src/test/java/java_baseball/java_baseball/ErrorTest.java (hunk 8 above)
Comment (shkisme) [reply shkisme/java_baseball/src/test/java/java_baseball/java_baseball/ErrorTest.java]: 따로 만든 ErrorMessage 클래스의 상수값을 이용하였습니다!

File: shkisme/java_baseball/src/test/java/java_baseball/java_baseball/RandomNumTest.java (hunk 9 above)
Comment (shkisme) [reply shkisme/java_baseball/src/test/java/java_baseball/java_baseball/RandomNumTest.java]: 아하 그렇군요! 배워갑니다. 랜덤 값에 대한 테스트라 랜덤 리스트의 크기 정도만 테스트 할 수 있는 것일까요...? 일단은 이렇게만 테스트 코드를 짰습니다.

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Constant.java (hunk 1 above)
Comment (shkisme) [reply shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Constant.java:3]: 확실히 모든 상수들을 한 클래스에 모아놓으니까 찾기도 어렵고 구분하기도 어렵더라구요... 각 상수가 쓰이는 클래스에 선언하였습니다! 감사합니당

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java (hunk 12 above)
Comment (shkisme) [reply shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java:10]: 저도 처음에 짤 때 Computer가 하는 역할이 좀 많나... 싶었던 것 같습니다. GameController를 만들어서 역할 분리를 했습니다!

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java (hunk 14 above)
Comment (shkisme) [reply shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java]: isGameEnd를 boolean으로 반환하니 gameState필드가 필요없어지군요😂 수정했습니다!

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Game.java (hunk 15 above)
Comment (shkisme) [reply shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Game.java]: 과거의 나는 왜 이렇게 짰던걸까요..? 수정했습니다!

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Player.java (hunk 16 above)
Comment (shkisme) [reply shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Player.java]: 각 함수의 역할에 따라 구분하여 분리했습니다!

File: shkisme/java_baseball/src/test/java/java_baseball/java_baseball/ErrorTest.java (hunk 17 above)
Comment (shkisme) [reply shkisme/java_baseball/src/test/java/java_baseball/java_baseball/ErrorTest.java]: edge case 테스트가 제일 중요할텐데 한쪽만 테스트했네요! 수정했습니다.

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java (hunk 19 above)
Comment (shkisme) [reply shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java]: 이름을 생각나는대로 적다보니 이런 실수를... 역할을 알 수 있도록 수정하였고, 각 역할에 따라 메서드를 분리했습니다!

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java [hunk 20]
```diff
@@ -0,0 +1,35 @@
+package java_baseball.java_baseball;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+class ComputerMessage{
```
Comment (gusah009) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java]: 한 파일에 `class`를 두 개를 두는 것보다 `ComputerMessage`클래스를 nested enum으로 넣는 게 더 깔끔할 것 같습니다!

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+import static java_baseball.java_baseball.RestartStatus.RESTART;
+
+class ErrorMessage {
+    public static final String LENGTH_ERROR = "숫자 3개를 입력해 주세요.";
+    public static final String VALUE_ERROR = "1부터 9 사이의 숫자를 입력해 주세요.";
+    public static final String DUPNUM_ERROR = "숫자를 중복 없이 입력해 주세요.";
+    public static final String RESTART_ERROR = "숫자 1 또는 2를 입력해 주세요.";
+}
+public class Error {
+    public void validateLengthError(String[] inputArr){
```
Comment (gusah009) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java]: 모든 메서드에 `Error`를 넣을 필요는 없을 것 같습니다! 

`Error` 클래스의 클래스명이 이를 얘기해주고 있어서 중복이라고 생각합니다.

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java [hunk 22]
```diff
@@ -0,0 +1,48 @@
+package java_baseball.java_baseball;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static java_baseball.java_baseball.RestartStatus.END;
+import static java_baseball.java_baseball.RestartStatus.RESTART;
+
+class ErrorMessage {
```
Comment (gusah009) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java]: > 한 파일에 class를 두 개를 두는 것보다 ComputerMessage클래스를 nested enum으로 넣는 게 더 깔끔할 것 같습니다!

위와 같은 내용입니다.

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+        }
+    }
+    public void validateNumValueError(List<Integer> playersNum){
+        for (int i= 0; i< Constant.LIST_SIZE;i++){
+            if (isValueError(playersNum.get(i))){
+                throw new IllegalArgumentException(ErrorMessage.VALUE_ERROR);
+            }
+        }
+    }
+    private boolean isValueError(int num){
```
Comment (gusah009) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java]: `is`를 붙이고 `boolean`을 반환하는 부분 잘 하신 것 같습니다! 👍

메서드 명만 조금 더 의미있게 써줘도 좋을 것 같습니다!

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+        if (Constant.BALL_MIN<=num && num <= Constant.BALL_MAX){
+            return false;
+        }
+        else{
+            return true;
+        }
+    }
+    public void validateDupNumError(List<Integer> playersNum){
```
Comment (gusah009) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java]: 긴 이름에 거부감을 느낄 걸 알지만, `Dup`으로 줄여쓰는 것 보다 `Duplicate`로 풀네임을 써주는 게 좋을 것 같습니다!

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/GameController.java [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+        this.player = player;
+        this.computer = computer;
+    }
+    public void countScore(){
+        saveStrikeCount(countStrike(player.playersNum));
+        saveBallCount(countBall(player.playersNum));
+    }
+    private int countStrike(List<Integer> playersNum){
+        int cnt =0;
+        for (int i= 0; i< Constant.LIST_SIZE; i++){
```
Comment (gusah009) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/GameController.java:25]: formatter를 쓰시는 걸 추천합니다! 아래는 KEEPER 홈페이지 백엔드 프로젝트에서 사용했던 formatter 입니다.

[홈페이지 백엔드 formatter](https://www.notion.so/921c1d8f25d940a594c686b7cf65b821#5e40164dc0214bff8a66b17b89f89350)

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/RestartStatus [hunk 26]
```diff
@@ -0,0 +1,15 @@
+package java_baseball.java_baseball;
+
+public enum RestartStatus {
```
Comment (gusah009) [inline shkisme/java_baseball/src/main/java/java_baseball/java_baseball/RestartStatus]: enum으로 잘 만든 것 같고 `getRestartStatus`를 두어서 캡슐화 한 것도 좋은 것 같습니다!

File: shkisme/java_baseball/src/test/java/java_baseball/java_baseball/ErrorTest.java [hunk 27]
```diff
@@ -0,0 +1,74 @@
...
+    @Test
+    @DisplayName("숫자를 3개보다 많이 입력한 경우")
+    public void lengthError(){
+        Constant constant = new Constant();
+        Error error = new Error();
+        String[] inputArr = {"4","5","2","1","7"};
+        IllegalArgumentException e = assertThrows(IllegalArgumentException.class,
+                () -> error.validateLengthError(inputArr));//예외가 발생해야 한다.
+
+        assertThat(e.getMessage()).isEqualTo(ErrorMessage.LENGTH_ERROR);
```
Comment (gusah009) [inline shkisme/java_baseball/src/test/java/java_baseball/java_baseball/ErrorTest.java]: 와우 코드리뷰 바로 반영한 코드 좋습니다~ 👍👍👍

File: shkisme/java_baseball/src/test/java/java_baseball/java_baseball/RestartTest [hunk 28]
```diff
@@ -0,0 +1,15 @@
+package java_baseball.java_baseball;
+
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import static java_baseball.java_baseball.RestartStatus.END;
+import static java_baseball.java_baseball.RestartStatus.RESTART;
+
+public class ReStartTest {
```
Comment (gusah009) [inline shkisme/java_baseball/src/test/java/java_baseball/java_baseball/RestartTest]: 몇몇 파일에 `.java` 확장자가 안붙어있는 것 같습니다! 😂

**Review by gusah009 — COMMENTED** [review]
코드리뷰 반영하시느라 고생 많으셨습니다!! 너무 잘하시네요 👍

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java (hunk 20 above)
Comment (shkisme) [reply shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Computer.java]: enum…을 많이 안써봐서 이렇게 구현이 잘 안되네요ㅜㅜ 수정했습니다!

File: shkisme/java_baseball/src/test/java/java_baseball/java_baseball/RestartTest (hunk 28 above)
Comment (shkisme) [reply shkisme/java_baseball/src/test/java/java_baseball/java_baseball/RestartTest]: 앗.. 수정했습니다!

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java (hunk 24 above)
Comment (shkisme) [reply shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java]: 메서드 이름을 자잘한 것까지 풀네임으로 수정했습니다!
메서드 이름 짓기 너무 난감해요😥

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/GameController.java (hunk 25 above)
Comment (shkisme) [reply shkisme/java_baseball/src/main/java/java_baseball/java_baseball/GameController.java:25]: 추천 감사합니다!

File: shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java (hunk 24 above)
Comment (gusah009) [reply shkisme/java_baseball/src/main/java/java_baseball/java_baseball/Error.java]: ㅋㅋㅋㅋㅋㅋㅋㅋㅋ 다른 사람들도 다 똑같나봐요

image

**Review by gusah009 — APPROVED** [review]
👍👍👍

File: shkisme/java_baseball/src/test/java/java_baseball/java_baseball/RandomNumTest.java (hunk 9 above)
Comment (woo-chang) [reply shkisme/java_baseball/src/test/java/java_baseball/java_baseball/RandomNumTest.java]: https://jojoldu.tistory.com/674

해당 글을 참고하면 제어할 수 없는 값은 테스트 하기 어렵다고 하네요 .. 각각의 메서드에서 랜덤적인 요소에 최대한 의존하지 않게 작성해보는 것도 좋을 것 같습니다!

**Review by woo-chang — APPROVED** [review]
진짜 피드백 수용이 너무 빠르셔서 감탄 .. 수고하셨습니다 👍👍

**Review by JeonghakLee — APPROVED** [review]
